### ISSUE TITLE
wishlist: be able to hide skipped roles when using conditional roles

### ISSUE THREAD

**Issue by jhagg** [ansible/ansible#9034]
##### Expected Results:

All roles where the expression with 'when' is false should not be shown
at all, preferrably by a config var and perhaps a way to reverse
on the command line.
##### Actual Results: task are always shown even if the role is skipped

I think conditional roles is a great idea, it made my plays easier to write.

However, all those skipped tasks are still shown which makes it harder to see the more interesting
output from the tasks that really are used.

Would it be possible to somehow hide them, maybe a config option in ansible.cfg that could
be reversed by a command line option?
(Yes, I realize that it is also very useful to see what actually is there even if it is not used :-)

**Comment by madduck** [comment]
cf. #3324

**Comment by riemers** [comment]
DISPLAY_SKIPPED_HOSTS=0 is a great addition, but it still shows output of things that don't do anything. I see this as "clutter" and makes the readability less on a playbook. I have several roles that show "Redhat" tasks, while i am using debian. No point in showing those.

**Comment by radudb** [comment]
+1

**Comment by janreges** [comment]
+1

In our Ansible installation we have about 50 roles. Our primary playbook defines roles conditionally by variables, which are defined per-host, or per-group.

Example playbook:

```
 hosts: "all"
  roles:
    - { role: siteone.apache, when: "enabled_role_siteone_apache == true"}
    - { role: siteone.backup, when: "enabled_role_siteone_backup == true"}
    - ...
```

This concept works very well for us, but it's very confusing, when only 1% of displayed tasks are relevant for my current host and 99% tasks are in "skipping" state.

**Comment by bennojoy** [comment]
@janreges  woudl this help
http://docs.ansible.com/intro_configuration.html#display-skipped-hosts we can set in ansible.cfg to not display skipped tasks

**Comment by eraserix** [comment]
+1

**Comment by konmeo** [comment]
+1

**Comment by ringerc** [comment]
This would make a huge difference when using conditionals. Because Ansible pushes conditionals on roles and on includes down into the tasks inside the role or include, something like:

```
- hosts: blah
  roles:
    - { role: my_server, my_server_cleanup_only: true }
```

can quickly produce huge spews of output for everything that the `my_server` role might do, but didn't do this time.

A possible workaround would initially seem to be a heirarchy of roles, where the base `my_server` role exposes the defaults, etc, then dependent roles like `my_server_cleanup` inherit it and perform particular steps. Then a `my_server_all` role that depends on all of them to do everything, so you don't have to enumerate a long list of roles for the common case. The problems with this are that it doesn't make life any easier when you want just a subset, and it repeats the base role execution for each and every time it gets inherited. Which is horrible, and much worse than the original problem.

Note that this is different to the `display_skipped_hosts` option. Without that option set:

```
TASK: [postgres_server | make clean] ****************************************** 
skipped: [nodeA07]
skipped: [nodeB07]
skipped: [nodeC07]

TASK: [postgres_server | make PostgreSQL] ************************************* 
skipped: [nodeA07]
skipped: [nodeB07]
skipped: [nodeC07]
```

With it:

```
TASK: [postgres_server | make clean] ****************************************** 

TASK: [postgres_server | make PostgreSQL] ************************************* 

```

... and what I actually want:

```
(nothing here)
```

**Comment by Shruti29** [comment]
+1

This feature request discussion has been going on since 2 years. 
It would be really useful if we have this implemented in the next version.

**Comment by lu4nm3** [comment]
+1

**Comment by c35sys** [comment]
+1

**Comment by oCroso** [comment]
+1 Please.

**Comment by denisbr** [comment]
+1
I tried looking into how to fix the code for this, but due to the way the TASK: bit is printed prior to checking (?) if the task should be skipped it was not trivial to create a patch for me with limited python skills.

**Comment by bcoca** [comment]
fixed via a09ce4bd8734421785b1e46f7bb15a15848d59dc, now you can use the 'skippy' callback to ignore all skips

**Comment by stucki** [comment]
Thanks for this. However, I'm not happy yet. Please see where the issue was continued from: #3324. It contains some examples where this makes sense, e.g. for registering variables with results from a shell command. For example, if I want to hide the output of a "grep" command, I may still want to see other skips. Therefore, skip and hide are not the same.

Besides this, I also think that this should be controllable per task and not just globally.

Question is, should I create a new (3rd) ticket for this, or is someone going to reopen here?

**Comment by bcoca** [comment]
They are different features, this ticket was to avoid displaying skipped tasks, #3324 was to hide tasks selectively. I don't see us implementing this currently, but you should be able to create your own callback plugin that does this.

**Comment by stucki** [comment]
Thanks for the feedback. Can you reopen #3324 then?

**Comment by bcoca** [comment]
why? I just stated I don't see us adding this feature.

**Comment by stucki** [comment]
The ticket was closed and even commenting was turned off without giving feedback to the various use cases that were brought up. I don't see why the issue was closed but esp. why the conversation was locked. Do you wish me to create a new issue for it?

**Comment by zerbons** [comment]
@bcoca would it be possible to port this to 1.9? If so what would need to be changed? Thanks!

**Comment by msabramo** [comment]
a09ce4b doesn't look terribly easy to use. I feel like with the number of +1's on #3324, it probably warrants introducing something that's built-in and easier to use.

**Comment by bcoca** [comment]
so setting `stdout_callback = skippy` in ansible.cfg is not easy enough?

**Comment by msabramo** [comment]
That looks pretty good actually but I didn't know about `stdout_callback` and I guess that's because it's a new thing in 2.0 and I don't see us moving to 2.0 for a long time.

Seems like in 1.9 I'd have to copy that file into my project's `plugins` directory? Or is there an easier way?

**Comment by bcoca** [comment]
no and copying it won't work, 2.0 pushes almost all output to the plugins, 1.x does not.

**Comment by solomongifford** [comment]
It appears that skippy in ansible.cfg does not suppress the output of the skipped roles/tasks, as was the original request. Instead it just suppresses the hosts within each task.

**Comment by varunkatta** [comment]
I ended up here googling "how to hide skipping messages in ansible". stdout_callback is not suppressing output of skipped tasks.  This is a desirable feature and should be supported.

**Comment by varunkatta** [comment]
I quickly hacked something to completely hide "skipped tasks" from stdout. It is not complete but it gives an idea on how to get this to work for your usecase. I haven't overridden all the callbacks but it seems to work as expected for the ones I did. 

For anyone who is interested use [this modified skippy plugin](https://github.com/varunkatta/ansible/commit/92d88c1571e4fa1792f7f4112d4ad609f59cace2) instead of the default one.

**Comment by hryamzik** [comment]
@varunkatta why don't you just implement this in a new plugin?

**Comment by fallspectrum** [comment]
+1

**Comment by asmartin** [comment]
+1

**Comment by s-sokolko** [comment]
+1

**Comment by tartley** [comment]
:-1

**Comment by tartley** [comment]
:-1:

**Comment by tartley** [comment]
I would really like this feature, more than any other new Ansible feature. Currently, if something goes wrong during a provision, I have 160kb of output (42 screensfull in a fullscreen small-font terminal) to visually parse to try and debug. Often, the relevant error, after lots of scrolling, is at the end. But sometimes it is not. This is regularly unnecessarily inconvenient.

I draw attention to the Unix philosophy's useful convention that a task which succeeds should produce no output (other than the data that was requested, if output is meant to be on stdout, which in Ansible's case it is not - we are running it for the side-effects.) Printing extra human-readable 'debug' output while running is almost always less useful than printing nothing except error messages.

I would assert that this is doubly true if the task in question is a no-op, because it is skipped.

https://en.wikipedia.org/wiki/Unix_philosophy (see 'rule of silence')

**Comment by hanej** [comment]
I tried using @varunkatta's plugin and it fails on Ansible 2.2.  This would be a great feature to have in Ansible and am quite surprised that it's not in there already.

**Comment by madduck** [comment]
Really, @hanej? I think you'll be up for some nasty surprises when you find out more about priorities and choices of this project…

**Comment by ringerc** [comment]
@madduck I know it's frustrating when it's not just what you want, but unless you're contributing considerable time or money to the project please appreciate what the team _are_ giving you, for free. I've been on the other side of this coin too, and it's incredibly demotivating to have people complaining at you as if they somehow inherently deserve more because what you did doesn't meet their specific use case.

Requests are fine. Suggestions are fine. Demands and complaints, not so much. Also, this is a bug tracker not a forum, so neither of us should really be digressing here, but I feel it's important to call this kind of comment out.

**Comment by madduck** [comment]
@ringerc I was a (early) contributor to this project before Michael shut me out because I was not 100% in line with his interests. Back then, I raised the issue that the template processing lost all data types in YAML and e.g. integers and booleans became text strings, which I considered a problem, but he didn't, and so that was it for me.

Anyway, this issue here (and the one I filed a long time before, #3324) are so popular that the way the project manager(s) (some of which are not volunteers) are handling this is indicative of something that I feel should not be ignored…

**Comment by bitdivine** [comment]
Piping the output into this gets rid of the skipped tasks, but then you lose the nice, shiny colours.  We could add the colours in again with another nice shiny bit of sed. ;-)

    ansible-playbook whatever.yml | sed -nr '/^TASK/{h;n;/^skipping:/{n;b};H;x};p'

**Comment by rbramwell** [comment]
I found this thread, looking for a similar solution. What worked for me was to set the following under the [defaults] section in my ansible.cfg file:

```bash
display_skipped_hosts = False
```

See for more info: http://docs.ansible.com/ansible/intro_configuration.html#display-skipped-hosts

**Comment by asmartin** [comment]
@rbramwell thanks for the suggestion, however even in this case the header is still printed (which takes up space and clutters the output). The ideal solution I am looking for is that only the tasks that are actually run are displayed (including the headers).

**Comment by senorsmile** [comment]
This has become an actual issue for us.  We have a huge role that skips a lot of things.  It usually only runs when a special variable is defined.  I just did a run against 29 hosts, and it took 0:04:57.888 according to the profiler to just skip the main task in that role.  That's 5 minutes of wasted scrolling of skipped tasks.  It is painful watching all the scrolling of skipped items for dozens of hosts.  We need a way to actually skip things.

**Comment by rangsan** [comment]
I have hundred of tasks that are skipped but still print out the task name which I have to scroll 100+ lines.
This request is really useful for us.

**Comment by jhawkesworth** [comment]
See https://github.com/ansible/ansible/pull/18621 - looks like the logic is there to do what you want but it would have to be a new plugin, or changed so that it is configurable in some way, rather than  modifying the existing behaviour of the 'skipply' callback plugin.

**Comment by brandonheller** [comment]
Putting the code from #18621 into callback_plugins/ and then referencing it in ansible_cfg like this:

```
[defaults]
stdout_callback = full_skip
```

...worked fine.  For releases past 2.4, it looks like the full_skip plugin will be available out-of-the-box.

**Comment by asmartin** [comment]
I tried putting the code from #18621 into `callback_plugins` and it works well, but if you're using `--diff` and a diff is printed, it is now printed ABOVE the task rather than below it. This is confusing since it seems that the diff is associated with the wrong task - can this be corrected so that the diff appears below the task as it normally would be?

**Comment by brandonheller** [comment]
@yasserg pointed out to me that full_skip.py doesn't print out the task name when it first executes, which is a problem when you're actively monitoring a task's execution and it isn't intended to complete immediately.  Many of our newer playbooks may have 90%+ skips, so this is a real issue for us; either we have to manually wade through everything, or we lose the ability to monitor jobs.

Looking at the code in https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py for Ansible's default strategy: determination of 'skipped' seems only available only after all hosts have executed that tasks (it's after line 296, where results are available).

So it seems like what I want might not even be possible as a plugin, unless you could defer the printout of the task start by a small amount of time, and if the playbook was skipped, nothing would print, but if the playbook was still executing, after say 100ms, it would print the task start and wait for other stuff.  And even that would require a hook executed after periodically as the task executed (or spawned as a separate thread).

Anyone else have ideas?  I'm no Ansible internals expert.  This is a real pain point for us.  Thanks!

**Comment by bcoca** [comment]
FYI, we dont normally 'see' closed tickets, but someone in irc asked me to weigh in on this.

@brandonheller actually the default callback already does this 'defered print' for the 'free' strategy, its a very old reporting trick, just need to keep track if you printed the header already , if not, print and set the flag that says you did.

**Comment by brandonheller** [comment]
Thanks @bcoca, I'll see if I can get that to work.  Regarding adding to a closed bug, I already saw a bunch of comments here after closure and figured that the context of the whole issue was relevant.  In the future should I create a new issue and link here?
